python/ray/serve/backend_state_manager.py

```diff
@@ -24,12 +24,51 @@
 SLOW_STARTUP_WARNING_PERIOD_S = 30
 
 
+class DeploymentState(Enum):
+    """
+    One abstraction level above replica to manage current state at
+    deployment level.
+
+    A deployment needs to keep track of two sets of states all time:
+      - (1) Target # of replicas running version X
+      - (2) Current # of replicas running version X
+
+    End user submits actions to change (1) through serve client,
+    such as:
+        - Deploying new deployment
+        - Add / remove replicas
+        - Update model version
+
+    Difference between (1) and (2) Determined current state of the
+    deployment that guides subsequent replica level actions.
+    """
+    # 0 replicas of the target state running
+    PENDING = 0
+    # Constructor failed N times for ALL replicas with 0 running,
+    # indicating last deployment action failed, will delete/rollback.
+    FAILED = 1
+    # > 1 but < target replica of the target state running, deployment
+    # is able to at least serve traffic
+    # All states below requires deployment to be at least in RUNNING
```

Comment:
Ah sorry for the confusion, i meant it should be only one at a time. Line 52 primarily means all subsequent states requires the deployment to at least reach RUNNING state first so deployment states have its own transition DAG and precedence.

Reply:
I see, makes sense. I think adding the `DeploymentState` here is fine but I'd prefer to avoid adding states that are not yet being used at all (in general I'm unilaterally opposed to checking in "dead code").